Part 1 intro to R (#5)
carpentries-incubator/r-geospatial-urban: episodes/03-explore-data.Rmd

---
title: 'Exploring Data Frames & Data frame Manipulation with dplyr '
teaching: 10
exercises: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE
)

# Load libraries ----------------------------------------------------------
# Package names
packages <- c("tidyverse", "here")

# Packages loading
invisible(lapply(packages, library, character.only = TRUE))

```

:::::::::::::::::::::::::::::::::::::: questions 

- What is a data frame?
- How can I read data in R?
- How can I get basic summary information about my data set?
- How can I select specific rows and/or columns from a data frame?
- How can I combine multiple commands into a single command?
- How can I create new columns or remove existing columns from a data frame?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

> - Describe what a data frame is.

Author: ```suggestion
After completing this episode, participants should be able to…

- Describe what a data frame is.
```

## Reading data

`read.csv()` is a function used to read coma separated data files (`.csv` format)). There are other functions for files separated with other delimiters. 
We're gonna read in the `gapminder` data set with information about countries' size, GDP and average life expectancy in different years.

```{r reading-data}
gapminder <- read_csv("data/gapminder_data.csv")

```

## Exploring dataset
Let’s investigate the `gapminder` data frame a bit; the first thing we should always do is check out what the data looks like.

It is important to see if all the variables (columns) have the data type that we require. For instance, a column might have numbers stored as characters, which would not allow us to make calculations with those numbers.

```{r inspecting-data-str}
str(gapminder) 

```

We can see that the `gapminder` object is a data.frame with `r nrow(gapminder)` observations (rows) and `r ncol(gapminder)` variables (columns). 

In each line after a `$` sign, we see the name of each column, its type and first few values. 

### First look at the dataset
There are multiple ways to explore a data set. Here are just a few examples:

```{r}

> head(gapminder) # see first 5  rows of the data set

Author: That's correct.

# [Data frame Manipulation with dplyr](https://datacarpentry.org/r-intro-geospatial/06-dplyr/index.html) 

## Select
Let's start manipulating the data. 

First, we will adapt our data set, by keeping only the columns we're interested in, using the `select()` function from the `dplyr` package:

```{r dplyr-select}
year_country_gdp <- select(gapminder, year, country, gdpPercap) 

head(year_country_gdp)

```

## Pipe
Now, this is not the most common notation when working with `dplyr` package. `dplyr` offers an operator `%>%` called a pipe, which allows you build up very complicated commands in a readable way.

In newer installation of `R` you can also find a notation `|>` . This pipe works in a similar way. The main difference is that you don't need to load any packages to have it available.

The `select()` statement with pipe would look like that:

```{r dplyr-pipe}

year_country_gdp <- gapminder %>% 
  select(year,country,gdpPercap)

head(year_country_gdp)

```

First we define data set, then - with the use of pipe we pass it on to the `select()` function. This way we can chain multiple functions together, which we will be doing now. 

## Filter

> We already know how to select only the needed columns. But now, we also want to filter the data set via certain conditions with `filter()` function. Instead of doing it in separate steps, we can do it all together. 

Reviewer: ```suggestion
We already know how to select only the needed columns. But now, we also want to filter the rows of our data set via certain conditions with `filter()` function. Instead of doing it in separate steps, we can do it all together. 
```

In the `gapminder` data set, we want to see the results from outside of Europe for the 21st century. 
```{r}
year_country_gdp_euro <- gapminder %>% 
  filter(continent != "Europe" & year >= 2000) %>% # & operator (AND) - both conditions must be met
  select(year, country, gdpPercap)

head(year_country_gdp_euro)
```

> Let's now find all the observations from Eurasia: 
> 
> ```{r}
> year_country_gdp_eurasia <- gapminder %>% 
>   filter(continent == "Europe" | continent == "Asia") %>% # I operator (OR) - one of the conditions must be met
>   select(year, country, gdpPercap)
> 
> head(year_country_gdp_eurasia)
> ```

Reviewer: ```suggestion

```

###  Exercise 1

<div class="alert alert-info">
<strong>Challenge</strong>

> Write a single command (which can span multiple lines and includes pipes) that will produce a dataframe that has the African values for life expectancy, country and year, but not for other Continents. How many rows does your data frame have and why?

Author: ```suggestion
Write a single command (which can span multiple lines and includes pipes) that will produce a data frame that has the values for life expectancy, country and year, only for Eurasia. How many rows does your data frame have and why? 
```

> Write a single command (which can span multiple lines and includes pipes) that will produce a data frame that has the values for life expectancy, country and year, only for Eurasia. How many rows does your data frame have and why? 

Author: An answer could be that the number of resulting rows is the cumulated number of countries in Europe and Asia.

<strong>Solution</strong>

</div>

```{r ex5, class.source="bg-info"}

> year_country_lifeExp_Africa <- gapminder %>%
>                            filter(continent=="Africa"  ) %>%
>                            select(year,country,lifeExp)
> 

Reviewer: ```suggestion
year_country_gdp_eurasia <- gapminder %>% 
  filter(continent == "Europe" | continent == "Asia") %>% # I operator (OR) - one of the conditions must be met
  select(year, country, gdpPercap)
```

Author: The operator is not `I` but the pipe `|`. I fixed the comment.

nrow(year_country_lifeExp_Africa)
``` 

## Group and summarize

> So far, we have created a data frame for one of the continents represented in the `gapminder` data set. But often instead of doing that, we would like to know statistics about all of the continents, presented by group.

Reviewer: ```suggestion
So far, we have provided summary statistics on the whole dataset or a particular subset. But often instead of doing that, we would like to know statistics about all continents, presented by group.
```

Reviewer: I had to change the introduction line as it did not fit anymore with the previous exercise

```{r dplyr-group}
gapminder %>% # select the dataset
  group_by(continent) %>% # group by continent
  summarize(avg_gdpPercap = mean(gdpPercap)) # summarize function creates statistics for the data set 

```

### Exercise 2
<div class="alert alert-info">
<strong>Challenge</strong> 
Calculate the average life expectancy per country. Which country has the longest average life expectancy and which has the shortest average life expectancy?

<strong>Hint</strong> Use `max()`  and `min()` functions to find minimum and maximum.

<strong>Solution</strong>

</div>

```{r ex6 , class.source="bg-info"}
gapminder %>%
   group_by(country) %>%
   summarize(avg_lifeExp=mean(lifeExp)) %>%
   filter(avg_lifeExp == min(avg_lifeExp) | avg_lifeExp == max(avg_lifeExp))

> 

Reviewer: we need to add ``` in this line

### Multiple groups and summary variables
You can also group by multiple columns:

```{r dplyr-group-multi}

gapminder %>%
  group_by(continent, year) %>%
  summarize(avg_gdpPercap = mean(gdpPercap))

```

On top of this, you can also make multiple summaries of those groups:
```{r dplyr-summ}
gdp_pop_bycontinents_byyear <- gapminder %>%
  group_by(continent,year) %>%
  summarize(
    avg_gdpPercap = mean(gdpPercap),
    sd_gdpPercap = sd(gdpPercap),
    avg_pop = mean(pop),
    sd_pop = sd(pop),
    n_obs = n()
    )

```

## Frequencies

If you need only a number of observations per group, you can use the `count()` function
```{r dplyr-count}

gapminder %>%
    group_by(continent) %>%
    count()
```
 

## Mutate

Frequently you’ll want to create new columns based on the values in existing columns. For example, instead of only having the GDP per capita, we might want to create a new GDP variable and convert its units into Billions. For this, we’ll use `mutate()`.

```{r dplyr-mutate}
gapminder_gdp <- gapminder %>%
  mutate(gdpBillion = gdpPercap*pop/10^9)

head(gapminder_gdp)

```

::::::::::::::::::::::::::::::::::::: keypoints 

> - 

Author: Looks good. I added some formatting.

> - 

Author: @fcjerome, @KyriJanssen, please add key points here. Please add them as suggestions via a comment so that we can commit the change directly.

Reviewer: I added some key points, however, they might be too very similar to the objectives stated at the beginning of the episode.

Author: Looks good.

# [Exploring Data frames](https://datacarpentry.org/r-intro-geospatial/04-data-structures-part2/index.html)

Now we turn to the bread-and-butter of working with `R`: working with tabular data. In `R` data are stored in a data structure called **data frames**.  

A data frame is a representation of data in the format of a **table** where the columns are **vectors** that all have the **same length**. 

Because columns are vectors, each column must contain a **single type of data** (e.g., characters, numeric, factors). 
For example, here is a figure depicting a data frame comprising a numeric, a character, and a logical vector.

![](assets/img/data-frame.svg)
<font size="3">*Source*:[Data Carpentry R for Social Scientists ](https://datacarpentry.org/r-socialsci/02-starting-with-data/index.html#what-are-data-frames-and-tibbles)</font>

## Reading data

`read.csv()` is a function used to read coma separated data files (`.csv` format)). There are other functions for files separated with other delimiters. 
We're gonna read in the `gapminder` data set with information about countries' size, GDP and average life expectancy in different years.

```{r reading-data}

> gapminder <- read.csv(here('data','gapminder_data.csv') )

Reviewer 2: why not use the dplyr function "read_csv"  for consistency?

Author: Done

In newer installation of `R` you can also find a notation `|>` . This pipe works in a similar way. The main difference is that you don't need to load any packages to have it available.

The `select()` statement with pipe would look like that:

```{r dplyr-pipe}

year_country_gdp <- gapminder %>% 
  select(year,country,gdpPercap)

head(year_country_gdp)

```

First we define data set, then - with the use of pipe we pass it on to the `select()` function. This way we can chain multiple functions together, which we will be doing now. 

## Filter

We already now how to select only the needed columns. But now, we also want to filter the data set via certain condition with `filter()` function. Instead doing it in separate steps, we can do it all together. 

In the `gapminder` data set, we want to see the results from outside of Europe for the 21st century. 
```{r}
year_country_gdp_euro <- gapminder %>% 
  filter(continent != "Europe" & year > 2000) %>% # & operator (AND) - both conditions must be met
  select(year, country, gdpPercap)

head(year_country_gdp_euro)
```

> Let's now find all the observations from Eurasia: 

Reviewer 2: Make this into a challenge?

Author: Good suggestion, @fcjerome. I made that change in the text of the next exercise.

```{r}
year_country_gdp_eurasia <- gapminder %>% 
  filter(continent == "Europe" | continent == "Asia") %>% # I operator (OR) - one of the conditions must be met
  select(year, country, gdpPercap)

head(year_country_gdp_eurasia)
```

###  Exercise 1

<div class="alert alert-info">
<strong>Challenge</strong>
Write a single command (which can span multiple lines and includes pipes) that will produce a dataframe that has the African values for life expectancy, country and year, but not for other Continents. How many rows does your data frame have and why?

> 

Author: We just used `filter()` and `select()` in the demonstration, so I don't think it is necessary to recall it in the challenge.

<strong>Solution</strong>

</div>

```{r ex5, class.source="bg-info"}
year_country_lifeExp_Africa <- gapminder %>%
                           filter(continent=="Africa"  ) %>%
                           select(year,country,lifeExp)

nrow(year_country_lifeExp_Africa)
``` 

## Group and summarize
So far, we have created a data frame for one of the continents represented in the `gapminder` data set. But often instead of doing that, we would like to know statistics about all of the continents, presented by group.

```{r dplyr-group}
gapminder %>% # select the dataset
  group_by(continent) %>% # group by continent
  summarize(avg_gdpPercap = mean(gdpPercap)) # summarize function creates statistics for the data set 

```

### Exercise 2
<div class="alert alert-info">
<strong>Challenge</strong> 

> Calculate the average life expectancy per country. Which country has the longest average life expectancy and which has the shortest average life expectancy?

Reviewer 2: 1 star for 2 good answers

<strong>Hint</strong> Use `max()`  and `min()` functions to find minimum and maximum.

<strong>Solution</strong>

</div>

```{r ex6 , class.source="bg-info"}
lifeExp_bycountry <- gapminder %>%
   group_by(country) %>%
   summarize(avg_lifeExp=mean(lifeExp))

lifeExp_bycountry %>%
   filter(avg_lifeExp == min(avg_lifeExp) | avg_lifeExp == max(avg_lifeExp))
```

### Multiple groups and summary variables
You can also group by multiple columns:

```{r dplyr-group-multi}

gapminder %>%
  group_by(continent, year) %>%
  summarize(avg_gdpPercap = mean(gdpPercap))

```

On top of this, you can also make multiple summaries of those groups:
```{r dplyr-summ}
gdp_pop_bycontinents_byyear <- gapminder %>%
  group_by(continent,year) %>%
  summarize(
    avg_gdpPercap = mean(gdpPercap),
    sd_gdpPercap = sd(gdpPercap),
    avg_pop = mean(pop),
    sd_pop = sd(pop),
    n_obs = n()
    )

```

## Frequencies

> If you need only a number of observations per group, you can use the `count()` function

Author: Good observation. Indeed, `count()` is a short version of that when `n()` is used. Good to mention it still.

```{r dplyr-count}

gapminder %>%
    group_by(continent) %>%
    count()
```
 

## Mutate

Frequently you’ll want to create new columns based on the values in existing columns, for example to do unit conversions, or to find the ratio of values in two columns. For this we’ll use `mutate()`.

```{r dplyr-mutate}
gapminder_gdp <- gapminder %>%
  mutate(gdpBillion = gdpPercap*pop/10^9)

head(gapminder_gdp)

```

> 

Reviewer 2: remove text below from Carpentry example

Reviewer 2: and add ggplot episode as ep04

Author: Done

Author: > remove text below from Carpentry example

Done

A data frame is a representation of data in the format of a **table** where the columns are **vectors** that all have the **same length**. 

Because columns are vectors, each column must contain a **single type of data** (e.g., characters, numeric, factors). 
For example, here is a figure depicting a data frame comprising a numeric, a character, and a logical vector.

![](assets/img/data-frame.svg)
<font size="3">*Source*:[Data Carpentry R for Social Scientists ](https://datacarpentry.org/r-socialsci/02-starting-with-data/index.html#what-are-data-frames-and-tibbles)</font>

## Reading data

`read.csv()` is a function used to read coma separated data files (`.csv` format)). There are other functions for files separated with other delimiters. 
We're gonna read in the `gapminder` data set with information about countries' size, GDP and average life expectancy in different years.

```{r reading-data}
gapminder <- read.csv("data/gapminder_data.csv")

```

## Exploring dataset
Let’s investigate the `gapminder` data frame a bit; the first thing we should always do is check out what the data looks like.

> It is important to see if all the variables (columns) have the data type that we require. Otherwise we can run into trouble.

Reviewer: 'Otherwise we can run into trouble' might be a bit vague. We could give an example of an issue we can encounter if we do not check the data types (e.g.,  trying to calculate the min, max, or mean of the variable that has been wrongly encoded as a character vector). However, this explanation might be too obvious.

Author: ```suggestion
It is important to see if all the variables (columns) have the data type that we require. For instance, a column might have numbers stored as characters, which would not allow us to make calculations with those numbers.
```

Author: Good suggestion, not obvious in this context.

</div>

```{r ex6 , class.source="bg-info"}
lifeExp_bycountry <- gapminder %>%
   group_by(country) %>%
   summarize(avg_lifeExp=mean(lifeExp))

lifeExp_bycountry %>%
   filter(avg_lifeExp == min(avg_lifeExp) | avg_lifeExp == max(avg_lifeExp))
```

### Multiple groups and summary variables
You can also group by multiple columns:

```{r dplyr-group-multi}

gapminder %>%
  group_by(continent, year) %>%
  summarize(avg_gdpPercap = mean(gdpPercap))

```

On top of this, you can also make multiple summaries of those groups:
```{r dplyr-summ}
gdp_pop_bycontinents_byyear <- gapminder %>%
  group_by(continent,year) %>%
  summarize(
    avg_gdpPercap = mean(gdpPercap),
    sd_gdpPercap = sd(gdpPercap),
    avg_pop = mean(pop),
    sd_pop = sd(pop),
    n_obs = n()
    )

```

## Frequencies

If you need only a number of observations per group, you can use the `count()` function
```{r dplyr-count}

gapminder %>%
    group_by(continent) %>%
    count()
```
 

## Mutate

Frequently you’ll want to create new columns based on the values in existing columns, for example to do unit conversions, or to find the ratio of values in two columns. For this we’ll use `mutate()`.

> 

Reviewer: Here, we can add a line explaining what we want to do in the next exercise.

e.g.
Frequently you’ll want to create new columns based on the values in existing columns. for example, instead of only having the GDP per capita, we might want to create a new  GDP variable and convert its units into Billions.

> Frequently you’ll want to create new columns based on the values in existing columns, for example to do unit conversions, or to find the ratio of values in two columns. For this we’ll use `mutate()`.

Reviewer: Here, we can add an explanation of what we want to do in the next exercise.
e.g.
Frequently you’ll want to create new columns based on the values in existing columns. for example, instead of only having the GDP per capita, we might want to create a new  GDP variable and convert its units into Billions. For this, we’ll use `mutate()`.

Author: ```suggestion
Frequently you’ll want to create new columns based on the values in existing columns. For example, instead of only having the GDP per capita, we might want to create a new GDP variable and convert its units into Billions. For this, we’ll use `mutate()`.
```
